src/borrow_tracker/tree_borrows/foreign_access_skipping.rs

```diff
@@ -0,0 +1,94 @@
+/// So this needs to be done manually, thus `ensure_no_stronger_than`, and also the test `permission_sifa_is_correct` in `perms.rs`
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Debug, Default)]
+pub enum LastForeignAccess {
+    #[default]
+    LocalAccess,
+    ForeignRead,
+    ForeignWrite,
+}
+
+impl LastForeignAccess {
+    /// Returns true if a node where the strongest idempotent foreign access is `self`
+    /// can skip the access `happening_next`. Note that if this returns
+    /// `true`, then the entire subtree will be skipped.
+    pub fn can_skip_foreign_access(self, happening_next: LastForeignAccess) -> bool {
+        debug_assert!(happening_next.is_foreign());
+        // This ordering is correct. Intuitively, if the last access here was
+        // a foreign write, everything can be skipped, since after a foreign write,
+        // all further foreign accesses are idempotent
+        happening_next <= self
+    }
+
+    /// Updates `self` to account for a foreign access.
+    pub fn record_new(&mut self, just_happened: LastForeignAccess) {
+        if just_happened.is_local() {
+            // If the access is local, reset it.
+            *self = LastForeignAccess::LocalAccess;
+        } else {
+            // Otherwise, keep it or stengthen it.
+            *self = just_happened.max(*self);
+        }
+    }
+
+    /// Returns true if this access is local.
+    pub fn is_local(self) -> bool {
+        matches!(self, LastForeignAccess::LocalAccess)
+    }
+
+    /// Returns true if this access is foreign, i.e. not local.
+    pub fn is_foreign(self) -> bool {
+        !self.is_local()
+    }
+
+    /// Constructs a foreign access from an `AccessKind`
+    pub fn foreign(acc: AccessKind) -> LastForeignAccess {
+        match acc {
+            AccessKind::Read => Self::ForeignRead,
+            AccessKind::Write => Self::ForeignWrite,
+        }
+    }
+
+    /// Usually, tree traversals have an `AccessKind` and an `AccessRelatedness`.
+    /// This methods converts these into the corresponding `LastForeignAccess`, to be used
+    /// to e.g. invoke `can_skip_foreign_access`.
+    pub fn of_acc_and_rel(acc: AccessKind, rel: AccessRelatedness) -> LastForeignAccess {
+        if rel.is_foreign() { Self::foreign(acc) } else { Self::LocalAccess }
+    }
+
+    /// During retags, the SIFA needs to be weakened to account for children with weaker SIFAs being inserted.
+    /// Thus, this method is called from the bottom up on each parent, until it returns false, which means the
+    /// "children have stronger SIFAs" invariant is restored.
+    pub fn ensure_no_stronger_than(&mut self, strongest_allowed: LastForeignAccess) -> bool {
+        if *self > strongest_allowed {
+            *self = strongest_allowed;
+            true
+        } else {
+            false
+        }
+    }
+}
+
+mod tests {
+    use super::LastForeignAccess;
```

Comment:
I get a warning during build that this `use` is not necessary. But if I remove it, I get an error:

```
error[E0433]: failed to resolve: use of undeclared type `LastForeignAccess`
  --> src/borrow_tracker/tree_borrows/foreign_access_skipping.rs:89:17
   |
89 |         assert!(LastForeignAccess::LocalAccess < LastForeignAccess::ForeignRead);
   |                 ^^^^^^^^^^^^^^^^^ use of undeclared type `LastForeignAccess`
   |
help: consider importing this enum
   |
87 +     use crate::borrow_tracker::tree_borrows::foreign_access_skipping::LastForeignAccess;
   |
```

(where again, it then tells me the inserted `use` is unused)